File: src/handlers/audioHandler.js
'use strict';

const { getMediaUrl, downloadMedia, sendMessage } = require('../services/whatsapp');
const { transcribeAudio } = require('../services/openai');
const { processExpenseText } = require('./textHandler');

/**
 * Maneja mensajes de audio (notas de voz).
 * Flujo: descarga → transcripción Whisper → parseo de gasto → confirmación
 *
 * @param {string} from
 * @param {{ id: string, mime_type: string }} audioMessage
 */
async function handleAudio(from, audioMessage) {
  await sendMessage(from, '🎤 Escuchando tu nota de voz...');

  let audioBuffer;
  try {
    const mediaUrl = await getMediaUrl(audioMessage.id);
    audioBuffer = await downloadMedia(mediaUrl);
  } catch (err) {
    console.error('[whatsapp] downloadMedia error:', err.message);
    await sendMessage(from, '❌ No pude descargar el audio. Intentá de nuevo o enviá un mensaje de texto.');
    return;
  }

  let transcription;
  try {
    transcription = await transcribeAudio(audioBuffer, audioMessage.mime_type);
  } catch (err) {
    console.error('[openai] transcribeAudio error:', err.message);
    await sendMessage(from, '❌ No pude transcribir el audio. Intentá enviar un mensaje de texto.');
    return;
  }

  if (!transcription) {
    await sendMessage(from, '❌ No entendí el audio. Intentá con un mensaje de texto.');
    return;
  }

  await sendMessage(from, `🎤 *Escuché:* "${transcription}"`);
  await processExpenseText(from, transcription);
}

module.exports = { handleAudio };

File: handlers/audio_handler.py
from services.whatsapp import send_message, get_media_url, download_media
from services.openai_service import transcribe_audio
from handlers.text_handler import process_expense_text


def handle_audio(phone: str, audio_msg: dict) -> None:
    """
    Descarga la nota de voz, la transcribe con Whisper y procesa el texto como gasto.
    audio_msg: dict con 'id' y 'mime_type' del mensaje de WhatsApp.
    """
    send_message(phone, "🎤 Escuchando tu nota de voz...")

    try:
        media_url = get_media_url(audio_msg["id"])
        audio_bytes = download_media(media_url)
    except Exception as e:
        print(f"[whatsapp] download_media error: {e}")
        send_message(
            phone,
            "❌ No pude descargar el audio. Intentá de nuevo o enviá un mensaje de texto.",
        )
        return

    try:
        transcript = transcribe_audio(audio_bytes, audio_msg.get("mime_type", "audio/ogg"))
    except Exception as e:
        print(f"[openai] transcribe_audio error: {e}")
        send_message(phone, "❌ No pude transcribir el audio. Intentá con un mensaje de texto.")
        return

    if not transcript:
        send_message(phone, "❌ No entendí el audio. Intentá con un mensaje de texto.")
        return

    send_message(phone, f'🎤 *Escuché:* "{transcript}"')
    process_expense_text(phone, transcript)
